### PR TITLE
Threading fix when populating the default library context.

### LINKED ISSUE #14720: Test failure
I'm seeing a thread sanitizer test failure:

```
WARNING: ThreadSanitizer: data race (pid=2647660)
  Write of size 4 at 0x7b2c000065b8 by main thread (mutexes: write M1624):
    #0 OPENSSL_LH_insert ../openssl/crypto/lhash/lhash.c:111 (libcrypto.so.3+0x2e1b9a)
    #1 lh_PROPERTY_DEFN_ELEM_insert ../openssl/crypto/property/defn_cache.c:30 (libcrypto.so.3+0x349ac7)
    #2 ossl_prop_defn_set ../openssl/crypto/property/defn_cache.c:112 (libcrypto.so.3+0x349f8c)
    #3 ossl_method_store_add ../openssl/crypto/property/property.c:255 (libcrypto.so.3+0x34aebb)
    #4 put_evp_method_in_store ../openssl/crypto/evp/evp_fetch.c:168 (libcrypto.so.3+0x29f7cb)
    #5 ossl_method_construct_this ../openssl/crypto/core_fetch.c:90 (libcrypto.so.3+0x2e5d2c)
    #6 algorithm_do_this ../openssl/crypto/core_algorithm.c:65 (libcrypto.so.3+0x2e5712)
    #7 ossl_provider_doall_activated ../openssl/crypto/provider_core.c:837 (libcrypto.so.3+0x2ff1dc)
    #8 ossl_algorithm_do_all ../openssl/crypto/core_algorithm.c:110 (libcrypto.so.3+0x2e5955)
    #9 ossl_method_construct ../openssl/crypto/core_fetch.c:124 (libcrypto.so.3+0x2e5fd2)
    #10 inner_evp_generic_fetch ../openssl/crypto/evp/evp_fetch.c:303 (libcrypto.so.3+0x29fe00)
    #11 evp_generic_fetch ../openssl/crypto/evp/evp_fetch.c:349 (libcrypto.so.3+0x2a006a)
    #12 EVP_MD_fetch ../openssl/crypto/evp/digest.c:1019 (libcrypto.so.3+0x274eee)
    #13 thread_multi_simple_fetch ../openssl/test/threadstest.c:351 (threadstest+0x579a)
    #14 test_multi ../openssl/test/threadstest.c:489 (threadstest+0x5f64)
    #15 run_tests ../openssl/test/testutil/driver.c:363 (threadstest+0xa998)
    #16 main ../openssl/test/testutil/main.c:30 (threadstest+0xb438)

  Previous atomic write of size 4 at 0x7b2c000065b8 by thread T11 (mutexes: write M1627):
    #0 __tsan_atomic32_store <null> (libtsan.so.0+0x7d82d)
    #1 OPENSSL_LH_retrieve ../openssl/crypto/lhash/lhash.c:171 (libcrypto.so.3+0x2e211c)
    #2 lh_PROPERTY_DEFN_ELEM_retrieve ../openssl/crypto/property/defn_cache.c:30 (libcrypto.so.3+0x349b4d)
    #3 ossl_prop_defn_get ../openssl/crypto/property/defn_cache.c:81 (libcrypto.so.3+0x349de9)
    #4 ossl_method_store_add ../openssl/crypto/property/property.c:251 (libcrypto.so.3+0x34ade4)
    #5 put_evp_method_in_store ../openssl/crypto/evp/evp_fetch.c:168 (libcrypto.so.3+0x29f7cb)
    #6 ossl_method_construct_this ../openssl/crypto/core_fetch.c:95 (libcrypto.so.3+0x2e5dfe)
    #7 algorithm_do_this ../openssl/crypto/core_algorithm.c:65 (libcrypto.so.3+0x2e5712)
    #8 ossl_provider_doall_activated ../openssl/crypto/provider_core.c:837 (libcrypto.so.3+0x2ff1dc)
    #9 ossl_algorithm_do_all ../openssl/crypto/core_algorithm.c:110 (libcrypto.so.3+0x2e5955)
    #10 ossl_method_construct ../openssl/crypto/core_fetch.c:124 (libcrypto.so.3+0x2e5fd2)
    #11 inner_evp_generic_fetch ../openssl/crypto/evp/evp_fetch.c:303 (libcrypto.so.3+0x29fe00)
    #12 evp_generic_fetch ../openssl/crypto/evp/evp_fetch.c:349 (libcrypto.so.3+0x2a006a)
    #13 EVP_MD_fetch ../openssl/crypto/evp/digest.c:1019 (libcrypto.so.3+0x274eee)
    #14 thread_multi_simple_fetch ../openssl/test/threadstest.c:351 (threadstest+0x579a)
    #15 thread_run ../openssl/test/threadstest.c:77 (threadstest+0x4359)
    #16 <null> <null> (libtsan.so.0+0x2d1af)

  Location is heap block of size 176 at 0x7b2c00006510 allocated by main thread:
    #0 malloc <null> (libtsan.so.0+0x30343)
    #1 CRYPTO_malloc ../openssl/crypto/mem.c:184 (libcrypto.so.3+0x2ece1f)
    #2 CRYPTO_zalloc ../openssl/crypto/mem.c:191 (libcrypto.so.3+0x2ece6b)
    #3 OPENSSL_LH_new ../openssl/crypto/lhash/lhash.c:51 (libcrypto.so.3+0x2e1836)
    #4 lh_PROPERTY_DEFN_ELEM_new ../openssl/crypto/property/defn_cache.c:30 (libcrypto.so.3+0x349a58)
    #5 property_defns_new ../openssl/crypto/property/defn_cache.c:61 (libcrypto.so.3+0x349d56)
    #6 ossl_lib_ctx_generic_new ../openssl/crypto/context.c:232 (libcrypto.so.3+0x2e4a7c)
    #7 ossl_crypto_alloc_ex_data_intern ../openssl/crypto/ex_data.c:432 (libcrypto.so.3+0x2ea3af)
    #8 ossl_lib_ctx_get_data ../openssl/crypto/context.c:336 (libcrypto.so.3+0x2e503f)
    #9 ossl_prop_defn_get ../openssl/crypto/property/defn_cache.c:74 (libcrypto.so.3+0x349db0)
    #10 ossl_method_store_add ../openssl/crypto/property/property.c:251 (libcrypto.so.3+0x34ade4)
    #11 put_evp_method_in_store ../openssl/crypto/evp/evp_fetch.c:168 (libcrypto.so.3+0x29f7cb)
    #12 ossl_method_construct_this ../openssl/crypto/core_fetch.c:90 (libcrypto.so.3+0x2e5d2c)
    #13 algorithm_do_this ../openssl/crypto/core_algorithm.c:65 (libcrypto.so.3+0x2e5712)
    #14 ossl_provider_doall_activated ../openssl/crypto/provider_core.c:837 (libcrypto.so.3+0x2ff1dc)
    #15 ossl_algorithm_do_all ../openssl/crypto/core_algorithm.c:110 (libcrypto.so.3+0x2e5955)
    #16 ossl_method_construct ../openssl/crypto/core_fetch.c:124 (libcrypto.so.3+0x2e5fd2)
    #17 inner_evp_generic_fetch ../openssl/crypto/evp/evp_fetch.c:303 (libcrypto.so.3+0x29fe00)
    #18 evp_generic_fetch ../openssl/crypto/evp/evp_fetch.c:349 (libcrypto.so.3+0x2a006a)
    #19 EVP_MD_fetch ../openssl/crypto/evp/digest.c:1019 (libcrypto.so.3+0x274eee)
    #20 thread_multi_simple_fetch ../openssl/test/threadstest.c:351 (threadstest+0x579a)
    #21 test_multi ../openssl/test/threadstest.c:489 (threadstest+0x5f64)
    #22 run_tests ../openssl/test/testutil/driver.c:363 (threadstest+0xa998)
    #23 main ../openssl/test/testutil/main.c:30 (threadstest+0xb438)

  Mutex M1624 (0x7b1000001e00) created at:
    #0 pthread_rwlock_init <null> (libtsan.so.0+0x3699f)
    #1 CRYPTO_THREAD_lock_new ../openssl/crypto/threads_pthread.c:41 (libcrypto.so.3+0x302186)
    #2 ossl_method_store_new ../openssl/crypto/property/property.c:186 (libcrypto.so.3+0x34aa6e)
    #3 evp_method_store_new ../openssl/crypto/evp/evp_fetch.c:33 (libcrypto.so.3+0x29f31c)
    #4 ossl_lib_ctx_generic_new ../openssl/crypto/context.c:232 (libcrypto.so.3+0x2e4a7c)
    #5 ossl_crypto_alloc_ex_data_intern ../openssl/crypto/ex_data.c:432 (libcrypto.so.3+0x2ea3af)
    #6 ossl_lib_ctx_get_data ../openssl/crypto/context.c:336 (libcrypto.so.3+0x2e503f)
    #7 get_evp_method_store ../openssl/crypto/evp/evp_fetch.c:75 (libcrypto.so.3+0x29f3d5)
    #8 inner_evp_generic_fetch ../openssl/crypto/evp/evp_fetch.c:228 (libcrypto.so.3+0x29f9af)
    #9 evp_generic_fetch ../openssl/crypto/evp/evp_fetch.c:349 (libcrypto.so.3+0x2a006a)
    #10 EVP_MD_fetch ../openssl/crypto/evp/digest.c:1019 (libcrypto.so.3+0x274eee)
    #11 thread_multi_simple_fetch ../openssl/test/threadstest.c:351 (threadstest+0x579a)
    #12 thread_run ../openssl/test/threadstest.c:77 (threadstest+0x4359)
    #13 <null> <null> (libtsan.so.0+0x2d1af)

  Mutex M1627 (0x7b1000004000) created at:
    #0 pthread_rwlock_init <null> (libtsan.so.0+0x3699f)
    #1 CRYPTO_THREAD_lock_new ../openssl/crypto/threads_pthread.c:41 (libcrypto.so.3+0x302186)
    #2 ossl_method_store_new ../openssl/crypto/property/property.c:186 (libcrypto.so.3+0x34aa6e)
    #3 alloc_tmp_evp_method_store ../openssl/crypto/evp/evp_fetch.c:64 (libcrypto.so.3+0x29f357)
    #4 ossl_method_construct ../openssl/crypto/core_fetch.c:116 (libcrypto.so.3+0x2e5efc)
    #5 inner_evp_generic_fetch ../openssl/crypto/evp/evp_fetch.c:303 (libcrypto.so.3+0x29fe00)
    #6 evp_generic_fetch ../openssl/crypto/evp/evp_fetch.c:349 (libcrypto.so.3+0x2a006a)
    #7 EVP_MD_fetch ../openssl/crypto/evp/digest.c:1019 (libcrypto.so.3+0x274eee)
    #8 thread_multi_simple_fetch ../openssl/test/threadstest.c:351 (threadstest+0x579a)
    #9 thread_run ../openssl/test/threadstest.c:77 (threadstest+0x4359)
    #10 <null> <null> (libtsan.so.0+0x2d1af)

  Thread T11 (tid=2647681, running) created by main thread at:
    #0 pthread_create <null> (libtsan.so.0+0x5ea99)
    #1 run_thread ../openssl/test/threadstest.c:83 (threadstest+0x43cc)
    #2 test_multi ../openssl/test/threadstest.c:486 (threadstest+0x5f34)
    #3 run_tests ../openssl/test/testutil/driver.c:363 (threadstest+0xa998)
    #4 main ../openssl/test/testutil/main.c:30 (threadstest+0xb438)

SUMMARY: ThreadSanitizer: data race ../openssl/crypto/lhash/lhash.c:111 in OPENSSL_LH_insert
==================
```

### PR DESCRIPTION
**Pull request by paulidale**
I'm not convinced that property definition locking is the right solution.
The test change fixes what was almost certainly a typo.

- [ ] documentation is added or updated
- [x] tests are added or updated

Fixes #14720

### CODE REVIEW

**Comment by XuanThuLe** [comment]
Ok

File: crypto/property/defn_cache.c [hunk 1]
```diff
@@ -98,25 +120,31 @@ int ossl_prop_defn_set(OSSL_LIB_CTX *ctx, const char *prop,
...
         }
-        if (!lh_PROPERTY_DEFN_ELEM_error(property_defns))
-            return 1;
+        if (!lh_PROPERTY_DEFN_ELEM_error(property_defns->defns))
+            goto fin;
     }
     OPENSSL_free(p);
-    return 0;
+    res = 0;
+ fin:
```
Comment (t8m) [inline crypto/property/defn_cache.c:147]: We usually use end, but OK. :grin:

Comment (paulidale) [reply crypto/property/defn_cache.c:147]: Changed.

**Comment by t8m** [comment]
This LGTM, but probably @levitte or @mattcaswell should look at it as well.

**Review by mattcaswell — APPROVED** [review]

**Comment by paulidale** [comment]
I'd still like to have a chat with @levitte about better fixes.  This entails not making the changes to `defn_cache.c` and doing something elsewhere.

**Comment by paulidale** [comment]
This needs reapproval because the dodgy add locking at the property definition level has been removed and in it's place I've added selective locking of the lib ctx.  The first two commits remain unchanged and are general improvements to the code.  The old third commit has been removed and replaced with a new leaner and better fix, one that locks the library context at just the right times.

**Review by mattcaswell — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to master, thanks for the feedback.

**Pull request closed without merging**